src/ptr/shared.zig

```diff
@@ -285,16 +285,13 @@ fn Weak(comptime Pointer: type, comptime options: Options) type {
 
         const SharedNonOptional = WithOptions(NonOptionalPointer, options);
 
-        /// Upgrades this weak pointer into a normal shared pointer.
-        ///
-        /// This method invalidates `self`.
-        pub fn upgrade(self: Self) ?SharedNonOptional {
+        /// Clones this weak pointer and upgrades it to a shared pointer.
+        pub fn cloneUpgrade(self: Self) ?SharedNonOptional {
```

Comment:
I agree that the “upgrade invalidates the weak pointer“ semantics were bad (an error on my part). Given that neither Rust's `Weak::upgrade` nor C++'s `weak_ptr::lock` destroy the weak pointer, I think we should just call this method `upgrade`, because there is no precedent for such an operation invalidating the weak pointer.

```suggestion
        /// Clones this weak pointer and upgrades it to a shared pointer. You still need to `deinit` the weak pointer.
        pub fn upgrade(self: Self) ?SharedNonOptional {
```